rfd/0175-static-host-users.md

```diff
@@ -0,0 +1,216 @@
+  rpc UpsertStaticHostUser(UpsertStaticHostUserRequest) returns (UpsertStaticHostUserResponse);
+  rpc DeleteStaticHostUser(DeleteStaticHostUserRequest) returns (google.protobuf.Empty);
+}
+
+message GetStaticHostUserRequest {
+    string name = 1;
+}
+
+message GetStaticHostUserResponse {
+    types.StaticHostUserV1 user = 1;
+}
+
+message ListStaticHostUsersRequest {
+    int32 page_size = 1;
+    string page_token = 2;
+}
+
+message ListStaticHostUsersResponse {
+    repeated types.StaticHostUserV1 users = 1;
+    string next_page_token = 2;
+}
+
+message CreateStaticHostUserRequest {
+    types.StaticHostUserV1 user = 1;
+}
+
+message CreateStaticHostUserResponse {
+    types.StaticHostUserV1 user = 1;
+}
+
+message UpdateStaticHostUserRequest {
+    types.StaticHostUserV1 user = 1;
+}
+
+message UpdateStaticHostUserResponse {
+    types.StaticHostUserV1 user = 1;
+}
+
+message UpsertStaticHostUserRequest {
+    types.StaticHostUserV1 user = 1;
+}
+
+message UpsertStaticHostUserResponse {
+    types.StaticHostUserV1 user = 1;
+}
+
+message DeleteStaticHostUserRequest {
+    string name = 1;
+}
+```
+
+### Propagation
+
+On startup, nodes will apply all available `static_host_user`s in the cache,
+then watch the cache for new and updated users. Nodes will use the labels in the
+`static_host_user`s to filter out those that don't apply to them, with the same
+logic that currently determines access with roles. Updated `static_host_user`s
+override the existing user. When a `static_host_user` is deleted, any host users
+created by it are *not* deleted (same behavior as `keep` mode for current host
+user creation).
+
+Nodes that disable host user creation (by setting `ssh_service.disable_create_host_user`
+to true in their config) will ignore `static_host_user`s entirely.
+
+### Audit events
+
+The `session.start` audit event will be extened to include a flag
+indicating whether or not the host user for an SSH session was
+created by Teleport (for both static and non-static host users).
+
+Two new audit events, `host_user.create` and `host_user.update`, will be added
+and emitted by nodes when they create or update a host user, respectively.
+
+### Product usage
+
+The session start PostHog event will be extended to include the
+same flag described in [Audit events](#audit-events).
+
+### Security
+
+CRUD operations on `static_host_user`s can be restricted with verbs
+in allow/deny rules like any other resource.
+
+We want to minimize the ability of Teleport users to mess with existing host users
+via `static_host_user`s. To that end, all host users created from `static_host_user`s
+will be in the `teleport-created` group (similar to the `teleport-system` group, which
+we currently use to mark users that Teleport should clean up). Teleport will not
+delete users not in `teleport-created`, and new users will not override existing users
+that are not in `teleport-created`.
```

Comment:
What about existing, potentially ephemeral users in `teleport-system`? Will they be adopted into `teleport-created`? If so we'd have to be careful about what happens right around the Teleport upgrade that introduces this feature, if the cluster is already configured for it. I think that we could make the case that if the user is not ephemeral then there's no real reason to override it - and if it is ephemeral, it will eventually get cleaned up by not having a session and then the static configuration can kick in "naturally".

Reply:
No, they won't be adopted. All users in `teleport-system` are ephemeral, so conflicts will be resolved after a session ends just as you described. In any case, I've renamed `teleport-created` to `teleport-static` to make it clearer that it's for marking static users, not all Teleport-created users.